src/OpenApi/src/Services/OpenApiDocumentService.cs

```diff
@@ -296,11 +296,98 @@ private OpenApiRequestBody GetFormRequestBody(IList<ApiRequestFormat> supportedR
             Content = new Dictionary<string, OpenApiMediaType>()
         };
 
-        // Forms are represented as objects with properties for each form field.
         var schema = new OpenApiSchema { Type = "object", Properties = new Dictionary<string, OpenApiSchema>() };
-        foreach (var parameter in formParameters)
+        // Group form parameters by their parameter name because MVC explodes form parameters that are bound from the
+        // same model instance into separate parameters, while minimal APIs does not.
+        //
+        // public record Todo(int Id, string Title, bool Completed, DateTime CreatedAt)
+        // public void PostMvc([FromForm] Todo person) { }
+        // app.MapGet("/form-todo", ([FromForm] Todo todo) => Results.Ok(todo));
+        //
+        // In the example above, MVC will bind four separate arguments to the Todo model while minimal APIs will
+        // bind a single Todo model instance to the todo parameter. Grouping by name allows us to handle both cases.
+        var groupedFormParameters = formParameters.GroupBy(parameter => parameter.ParameterDescriptor.Name);
+        // If there is only one real parameter derived from the form body, then set it directly in the schema.
+        var hasMultipleFormParameters = groupedFormParameters.Count() > 1;
+        foreach (var parameter in groupedFormParameters)
         {
-            schema.Properties[parameter.Name] = _componentService.GetOrCreateSchema(parameter.Type);
+            if (parameter.All(parameter => parameter.ModelMetadata.ContainerType is null))
+            {
+                var description = parameter.Single();
+                // Form files are keyed by their parameter name so we must capture the parameter name
+                // as a property in the schema.
+                if (description.Type == typeof(IFormFile) || description.Type == typeof(IFormFileCollection))
+                {
+                    if (hasMultipleFormParameters)
+                    {
+                        schema.AllOf.Add(new OpenApiSchema
+                        {
+                            Type = "object",
+                            Properties = new Dictionary<string, OpenApiSchema>
+                            {
+                                [description.Name] = _componentService.GetOrCreateSchema(description.Type)
+                            }
+                        });
+                    }
+                    else
+                    {
+                        schema.Properties[description.Name] = _componentService.GetOrCreateSchema(description.Type);
+                    }
+                }
+                else
+                {
+                    if (hasMultipleFormParameters)
+                    {
+                        if (description.ModelMetadata.IsComplexType)
```

Comment:
This is pulled out because we can refer to the complex type by name and don't have to create an inline schema?  Assuming so, is it possible for the complex type to be anonymous or unnameable?

Reply:
All the schemas that we generate are currently inline at the moment. This is more-so about the fact that schemas for complex types don't need to be nested under the parameter name based on the way parameters are bound from the form.

For an endpoint like:

```
app.MapPost("/", ([FromForm] Todo todo => { });
```

We want to generate the following schema:

```json
{
	"properties": {
		"id":  { ... },
		"title": { ... },
		"isCompleted": { ... }
	}
}
```

Not this one:

```json
{
	"properties": {
		"todo": {
			"properties": {
				"id":  { ... },
				"title": { ... },
				"isCompleted": { ... }
			}
		}
	}
}
```